#2823 update form date input date parsing

wlthomson opened the pull request:
Fixes #2823.

## Change summary

Fixes bug in validation logic which was causing the search button to be disabled when the date of birth field was reset.

## Testing

Steps to reproduce or otherwise test the changes of this PR:

- [ ] The patient lookup search button is not disabled when the modal is opened.
- [ ] The patient lookup search button is not disabled when the date of birth field is updated.
- [ ] The patient lookup search button is not disabled when the date of birth field is reset to blank.

### Related areas to think about

N/A.

mark-prins commented:
when I enter a date and search, I get this every time. Checking now - just making a note to alert other potential reviewers

![Screen Shot 2020-05-08 at 3 44 55 PM](https://user-images.githubusercontent.com/9192912/81368094-0a4ee380-9143-11ea-9e93-77ede34b4490.png)

wlthomson commented:
@mark-prins That's odd, I thought that was the case and changed it after getting that error. I must have got it backwards!

Update:

The `onUpdate` date formatting wasn't the issue (it works either way). The problem was the server rejecting the request with a date in the query string, as the 4D routing doesn't handle backslashes in query strings. The routing was splitting on all `/` characters to split the URL into components, even if the `/` was escaped and came after the `&` separator. 

I've fixed the query string constructor, so now date queries are working fine (for me). 

Note: if you're getting `Invalid date` records that don't match, make sure your structure has this fix: https://github.com/sussol/msupply/commit/c4d5bf2f288588f2252341552e6f392a15a8b35a.

diegoevangelisti commented:
> when I enter a date and search, I get this every time. Checking now - just making a note to alert other potential reviewers
> 
> ![Screen Shot 2020-05-08 at 3 44 55 PM](https://user-images.githubusercontent.com/9192912/81368094-0a4ee380-9143-11ea-9e93-77ede34b4490.png)

@wlthomson @mark-prins I am getting the same here - `No records found` when trying to match an existing Patient entering a First Name, Last Name and DOB. (It does work with just First and Last Name).

mark-prins commented on src/sync/lookupApiUtils.js:
```diff
@@ -37,20 +45,22 @@ export const createPolicyRecord = policy => {
+    const paramValue = type !== TYPES.DATE ? `@${value}@` : moment(value).format('YYYYMMDD');
+    return { ...queryObject, [key]: paramValue };
+  }, {});
+  const queryString = `?${querystring.stringify(object)}`;
+  return queryString;
```
mainly because the mix of 'querystring' and 'queryString' tripped me up.
```suggestion
  return `?${querystring.stringify(object)}`;
```

wlthomson replied:
49ca9cc.

mark-prins commented on src/sync/lookupApiUtils.js:
```diff
@@ -37,20 +45,22 @@ export const createPolicyRecord = policy => {
+  const object = params.reduce((queryObject, param) => {
+    const [[key, value], [, type]] = Object.entries(param);
+    if (!value) return queryObject;
+    const paramValue = type !== TYPES.DATE ? `@${value}@` : moment(value).format('YYYYMMDD');
```
The date format should be `DDMMYYYY`

<img width="393" alt="Screen Shot 2020-05-11 at 1 23 06 PM" src="https://user-images.githubusercontent.com/9192912/81516203-38beff80-938b-11ea-8f7e-40f583e9959a.png">

wlthomson replied:
![oops](https://user-images.githubusercontent.com/43223637/81528459-d678f580-93b0-11ea-9d2b-d938df92bf08.gif)

49ca9ccc1f65540776b674b0eeeb5d392fc27b46.

mark-prins replied:
![dobby-finger-snap](https://user-images.githubusercontent.com/9192912/81614275-e8e84300-9433-11ea-90df-3cd18af0d947.gif)

( works a treat! thanks )

wlthomson commented:
@diegoevangelisti Date not matching would have been due to the format inconsistency in @mark-prins review. Should work now.
